### PR TITLE
add blog about doing translations in vue

### PR DESCRIPTION
**Pull request by hillalex**
Add a new blog post and also adds Naomi to the list of projects.

### CODE REVIEW

File: content/blog/vue-translate.md [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+We've spent the last few months developing [Naomi](/projects/#naomi) - a web app for interfacing with
+ an HIV model developed by researchers at GIDA in association with UNAIDS. To support the West African
+ countries who are the users of Naomi we are providing two language modes - French and English. 
+ 
+ The front-end of the app is written in [Vue.js](https://vuejs.org/) and makes use of [Vuex](https://vuex.vuejs.org/),
+ a [flux](https://facebook.github.io/flux/) library for Vue. There are two ways we could have implemented
+ translations in Vue - with components or directives. We opted for the latter; I'll explain why and how!
+
+### Components vs directives
+Vue, like React, is a framework for writing components.
```
Comment (richfitz) [inline content/blog/vue-translate.md]: could you link React here?

File: content/blog/vue-translate.md [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+### Components vs directives
+Vue, like React, is a framework for writing components. 
+A component looks like a custom HTML element, but has special behaviour attached to it by Vue.
+Vuex provides helper functions that allow
+components to subscribe to changes in the store, so it's pretty straight-forward to define a component that will
+ perform translations. Usage of such a component would look a little like this:
+
+```
+<h3><translated translation-key="welcomeMessage"></translated></h3>
```
Comment (richfitz) [inline content/blog/vue-translate.md:23]: What do you think about adding a tiny json snippet showing the key-value pairs for two languages?

File: content/projects.md [hunk 3]
```diff
@@ -17,13 +17,24 @@ The [Vaccine Impact Modelling Consortium](https://www.vaccineimpact.org/) coordi
...
 
 * [Paper describing the model](https://www.biorxiv.org/content/10.1101/532010v1)
 
+### Naomi
+[https://naomi.unaids.org](https://naomi.unaids.org/)
+
+A web interface for a model estimating various HIV epidemic indicators at a regional level. For use by 
+countries in association with UNAIDS. The app is comprised of:
+
+* an R API and R workers, managed by a Redis queue
```
Comment (richfitz) [inline content/projects.md]: Can you add links to the components?

File: content/blog/vue-translate.md [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+        language: "en     
+    } 
+})
+```
+
+For performing the translations we have used [i18next](https://www.i18next.com/), which is intialised
+ with a language and the available translations. When the user selects a new language we fire off a Vuex
+  action that updates both i18next and the language in our store state.
+ 
+ ```
```
Comment (richfitz) [inline content/blog/vue-translate.md]: It is no clear to me (as someone who knows nothing about vue) where this code sits? Does that even matter?  This is obviously in your js code somewhere!

File: content/blog/vue-translate.md [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+Vue.directive('translate', translate);
+```
+
+The above is the code for a directive that just translates text. The directive we created is able to 
+translate arbitrary attributes of an element as per:
+
+```<input v-translate:placeholder="'email'" />```
+
+You can see the full (typescript) code for the final directive
+ [here](https://github.com/mrc-ide/hint/blob/master/src/app/static/src/app/directives/translate.ts).
```
Comment (richfitz) [inline content/blog/vue-translate.md:226]: I sent looking for the `changeLanguage` bit and could not find that in here (as earlier in the post) but found it in the store/root/actions.ts file - is that worth linking here too?

File: content/blog/vue-translate.md [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+  action that updates both i18next and the language in our store state.
+ 
+ ```
+ await i18next.changeLanguage(newLanguage);
+ commit({type: "CHANGE_LANGUAGE", payload: newLanguage})
+```
+
+### Listening for changes
+
+Listening for changes in a component is easy, it's baked into the framework. A simple translation component
```
Comment (richfitz) [inline content/blog/vue-translate.md]: Possibly worth adding something for naive people (like me) that this is *not* what you did, as I'd forgotten that we're not doing a component but instead a directive and got confused trying to relate this to your translate.ts implementation.

**Review by richfitz — COMMENTED** [review]
If you had the appetite to set up a jsfiddle type thing that showed it all off that'd be super cool. I just about follow what is going on, but I don't think I've got the required level of js.

The i18next page lists three plugins (https://github.com/rse/vue-i18next, https://github.com/panter/vue-i18next, https://github.com/bluelovers/vue-i18next2) - were these not appropriate because they did not support the flux model? Are they worth mentioning?

Thanks for adding naomi to the the project page!

File: content/blog/vue-translate.md [hunk 7]
```diff
@@ -0,0 +1,175 @@
+---
+author: "Alex Hill"
+date: 2019-12-17
+title: A custom directive for translations in Vue.js
+---
+
+We've spent the last few months developing [Naomi](/projects/#naomi) - a web app for interfacing with
+ an HIV model developed by researchers at GIDA in association with UNAIDS. To support the West African
```
Comment (EmmaLRussell) [inline content/blog/vue-translate.md]: Links to both orgs would be nice.

File: content/blog/vue-translate.md [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+like a custom attribute on an HTML element and defines some behaviour on that element. It would be much
+less intrusive if we could define a custom directive that would translate any part of an existing element.
+The desired interface would be:
+
+```
+<h3 v-translate="'welcomeMessage'"></h3>
+<input type="text" v-translate:placeholder="'email'" name="myEmailInput" value="" />
+```
+
+This is clearly a much nicer interface, but is not as straight-forward to implement. This blog will describe
```
Comment (EmmaLRussell) [inline content/blog/vue-translate.md]: If it isn't obvious to the reader why this is a nicer interface, could maybe explain parallels with built-in vue directives like v-text to define component behaviour, while keeping the generated HTML clean and not proliferating elements/components.

File: content/blog/vue-translate.md [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+<translated-input 
+            translation-key="email" 
+            translation-attribute="placeholder" 
+            inputType="text"
+            name="myEmailInput"
+            value="">
+</translated-input>
+```
+
+But unlike React and more like Angular, Vue is also a framework of directives. A directive looks
```
Comment (EmmaLRussell) [inline content/blog/vue-translate.md]: If there's going to be a link to React, Angular probably deserves one too...

File: content/blog/vue-translate.md [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+    props: ["title"]
+}
+</script>
+```
+
+But there is no directive hook that will fire when *store state* changes, i.e. our language. So we'll have to manage
+a subscription to the store ourselves. We can set up such a subscription on `bind`:
+
+```
+bind(el, binding) {
```
Comment (EmmaLRussell) [inline content/blog/vue-translate.md:177]: Maybe put in more of an explanation of the getting and using of the unwatch function since that confused me!

**Review by EmmaLRussell — COMMENTED** [review]
Nice, seems very clear to me.

File: content/blog/vue-translate.md [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+how we did it!
+
+### Storing and updating the user's chosen language
+In Vuex all app state is held
+in a central "store", which is then the ideal place to keep track of what language the user has selected.
+
+```
+const store = new Vuex.Store({
+    state: {
+        language: "en
```
Comment (giovannic) [inline content/blog/vue-translate.md]: ending quote missing here?

Comment (giovannic) [inline content/blog/vue-translate.md]: missing closing quote here?

File: content/blog/vue-translate.md (hunk 5 above)
Comment (hillalex) [reply content/blog/vue-translate.md:226]: It's a bit convoluted in our app now, so I've added a more complete code snippet above instead

**Comment by hillalex** [comment]
> If you had the appetite to set up a jsfiddle type thing that showed it all off that'd be super cool. I just about follow what is going on, but I don't think I've got the required level of js.
> 
> The i18next page lists three plugins (https://github.com/rse/vue-i18next, https://github.com/panter/vue-i18next, https://github.com/bluelovers/vue-i18next2) - were these not appropriate because they did not support the flux model? Are they worth mentioning?
> 
> Thanks for adding naomi to the the project page!

Those plugins all had quite strange interfaces and none of them worked with Vuex out of the box. I spent some time reading the documentation for the best documented one and I couldn't really understand what it was adding over and above `i81next` itself!

Added a fiddle :)

**Review by richfitz — APPROVED** [review]
This is great :) - the fiddle is particularly cool

You should probably update the date to reflect when you want it published as it's currently 2 days old

**Pull request merged**